packages/framework/client-logger/external-telemetry/src/container/telemetryProducer.ts

```diff
@@ -0,0 +1,107 @@
+ */
+
+import {
+	IContainer,
+	IContainerEvents,
+	type ICriticalContainerError,
+} from "@fluidframework/container-definitions";
+import {
+	ContainerConnectedTelemetry,
+	ContainerTelemetryEventNames,
+	IContainerTelemetry,
+	ContainerDisconnectedTelemetry,
+	ContainerClosedTelemetry,
+	ContainerAttachingTelemetry,
+	ContainerAttachedTelemetry,
+	type ContainerTelemetryEventName,
+} from "./containerTelemetry";
+import { ContainerSystemEventName, ContainerSystemEventNames } from "./containerSystemEvents";
+
+/**
+ * This class produces {@link IContainerTelemetry} from raw container system events {@link IContainerEvents}.
+ * The class contains different helper methods for simplifying and standardizing logic for adding additional information necessary
+ * to produce different {@link IContainerTelemetry}.
+ *
+ * @internal
+ */
+export class ContainerEventTelemetryProducer {
+	constructor(private container: IContainer) {}
+
+	public produceTelemetry(
+		eventName: ContainerSystemEventName,
+		payload?: any,
+	): IContainerTelemetry | undefined {
+		let telemetry: IContainerTelemetry | undefined = undefined;
+		switch (eventName) {
+			case ContainerSystemEventNames.CONNECTED:
+				telemetry = this.produceConnectedTelemetry(payload);
+				return telemetry;
+			case ContainerSystemEventNames.DISCONNECTED:
+				telemetry = <ContainerDisconnectedTelemetry>(
+					this.produceBasicContainerTelemetry(ContainerTelemetryEventNames.DISCONNECTED)
+				);
+				break;
+			case ContainerSystemEventNames.CLOSED:
+				telemetry = this.produceClosedTelemetry(payload);
+				break;
+			case ContainerSystemEventNames.ATTACHED:
+				telemetry = <ContainerAttachedTelemetry>(
+					this.produceBasicContainerTelemetry(ContainerTelemetryEventNames.ATTACHED)
+				);
+				break;
+			case ContainerSystemEventNames.ATTACHING:
+				telemetry = <ContainerAttachingTelemetry>(
+					this.produceBasicContainerTelemetry(ContainerTelemetryEventNames.ATTACHING)
+				);
+				break;
+			default:
+				break;
+		}
+		return telemetry;
+	}
+
+	private produceBasicContainerTelemetry = <T extends IContainerTelemetry>(
+		eventName: ContainerTelemetryEventName,
+	): T => {
+		return {
+			eventName,
+			containerId: this.getContainerId(),
+			documentId: this.getDocumentId(),
+		} as T;
+	};
+
+	private produceConnectedTelemetry = (payload?: {
+		clientId: string;
+	}): ContainerConnectedTelemetry => {
+		return {
+			eventName: ContainerTelemetryEventNames.CONNECTED,
+			containerId: payload?.clientId ?? this.getContainerId(),
+			documentId: this.getDocumentId(),
+		};
+	};
+
+	private produceClosedTelemetry = (payload?: {
+		error?: ICriticalContainerError;
+	}): ContainerClosedTelemetry => {
+		const telemetry: ContainerClosedTelemetry = {
+			eventName: ContainerTelemetryEventNames.CLOSED,
+			containerId: this.getContainerId(),
+			documentId: this.getDocumentId(),
+		};
+		if (payload?.error !== undefined) {
+			telemetry.error = payload.error;
+		}
```

Comment:
I think this accomplishes the same thing? Even if the it makes a difference where AppInsights ends up having an error column with no values VS no error column, I think that's fine.
```suggestion
		const telemetry: ContainerClosedTelemetry = {
			eventName: ContainerTelemetryEventNames.CLOSED,
			containerId: this.getContainerId(),
			documentId: this.getDocumentId(),
			error: payload?.error
		};
```

Reply:
Discussed offline - this is to keep consistency with how an object with a key has an undefined attribute show up. An end destination for the telemetry may have logic that show attributes with an undefined value, or it may ignore them entirely so with that in mind we can control the output the approach here.